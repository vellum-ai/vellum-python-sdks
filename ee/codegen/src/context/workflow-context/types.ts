export interface SDK_MODULE_PATHS {
  WORKFLOWS_MODULE_PATH: readonly string[];
  CORE_NODES_MODULE_PATH: readonly string[];
  DISPLAYABLE_NODES_MODULE_PATH: readonly string[];
  INPUTS_MODULE_PATH: readonly string[];
  SANDBOX_RUNNER_MODULE_PATH: readonly string[];
  STATE_MODULE_PATH: readonly string[];
  NODE_DISPLAY_MODULE_PATH: readonly string[];
  NODE_DISPLAY_TYPES_MODULE_PATH: readonly string[];
  VELLUM_TYPES_MODULE_PATH: readonly string[];
  PORTS_MODULE_PATH: readonly string[];
}
