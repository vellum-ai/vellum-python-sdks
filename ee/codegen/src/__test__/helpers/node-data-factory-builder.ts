import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
