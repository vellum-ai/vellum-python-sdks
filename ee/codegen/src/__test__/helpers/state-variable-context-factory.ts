import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
