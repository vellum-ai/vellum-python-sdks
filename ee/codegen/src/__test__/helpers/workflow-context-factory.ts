import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
