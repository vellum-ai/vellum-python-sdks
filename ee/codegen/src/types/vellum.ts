import {
  ChatMessage,
  ConditionCombinator,
  LogicalOperator,
} from "vellum-ai/api";
import {
  ChatHistoryInput,
  ChatMessageRequest,
  ChatMessageRole,
  JsonInput,
  NumberInput,
  PromptBlockState,
  PromptParameters,
  SearchResult,
  SearchResultRequest,
  StringInput,
  VellumValue,
  VellumVariable,
  VellumVariableType,
} from "vellum-ai/api/types";

export enum WorkflowNodeType {
  PROMPT = "PROMPT",
  TEMPLATING = "TEMPLATING",
  NOTE = "NOTE",
  CODE_EXECUTION = "CODE_EXECUTION",
  METRIC = "METRIC",
  SEARCH = "SEARCH",
  MERGE = "MERGE",
  CONDITIONAL = "CONDITIONAL",
  API = "API",
  ENTRYPOINT = "ENTRYPOINT",
  TERMINAL = "TERMINAL",
  SUBWORKFLOW = "SUBWORKFLOW",
  MAP = "MAP",
  ERROR = "ERROR",
  GENERIC = "GENERIC",
}

export enum ConditionalCombinator {
  OR = "OR",
  AND = "AND",
}

export interface StringVellumValue {
  type: "STRING";
  value: string;
}

export interface NumberVellumValue {
  type: "NUMBER";
  value: number;
}

export interface ChatHistoryVellumValue {
  type: "CHAT_HISTORY";
  value: ChatMessageRequest[] | ChatMessage[];
}

export interface SearchResultsVellumValue {
  type: "SEARCH_RESULTS";
  value: SearchResultRequest[] | SearchResult[];
}

export interface JsonVellumValue {
  type: "JSON";
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  value?: any;
}

export interface ConstantValuePointer {
  type: "CONSTANT_VALUE";
  data: VellumValue;
}

export type VellumLogicalExpression =
  | VellumLogicalCondition
  | VellumLogicalConditionGroup;

export interface VellumLogicalCondition {
  type: "LOGICAL_CONDITION";
  lhsVariableId: string;
  operator: LogicalOperator;
  rhsVariableId: string;
}

export interface VellumLogicalConditionGroup {
  type: "LOGICAL_CONDITION_GROUP";
  conditions: VellumLogicalExpression[];
  combinator: ConditionCombinator;
  negated: boolean;
}

export interface NodeOutputData {
  nodeId: string;
  outputId: string;
}

export interface NodeOutputPointer {
  type: "NODE_OUTPUT";
  data: NodeOutputData;
}

export interface InputVariableData {
  inputVariableId: string;
}

export interface InputVariablePointer {
  type: "INPUT_VARIABLE";
  data: InputVariableData;
}

export interface WorkspaceSecretData {
  type: "STRING";
  workspaceSecretId?: string;
}

export interface WorkspaceSecretPointer {
  type: "WORKSPACE_SECRET";
  data: WorkspaceSecretData;
}

export interface ExecutionCounterData {
  nodeId: string;
}

export interface ExecutionCounterPointer {
  type: "EXECUTION_COUNTER";
  data: ExecutionCounterData;
}

export type NodeInputValuePointerRule =
  | NodeOutputPointer
  | InputVariablePointer
  | ConstantValuePointer
  | WorkspaceSecretPointer
  | ExecutionCounterPointer;

export interface NodeInputValuePointer {
  rules: NodeInputValuePointerRule[];
  combinator: "OR";
}

export interface NodeInput {
  id: string;
  key: string;
  value: NodeInputValuePointer;
}

export interface NodeDisplayPosition {
  x: number;
  y: number;
}

export interface NodeDisplayComment {
  value?: string;
  expanded?: boolean;
}

export interface NodeDisplayData {
  position?: NodeDisplayPosition;
  width?: number;
  height?: number;
  comment?: NodeDisplayComment;
}

export interface CodeResourceDefinition {
  name: string;
  module: string[];
}

export type MergeBehavior =
  | "AWAIT_ATTRIBUTES"
  | "AWAIT_ALL"
  | "AWAIT_ANY"
  | "CUSTOM";

export interface NodeTrigger {
  id: string;
  mergeBehavior: MergeBehavior;
}

export interface DefaultNodePort {
  type: "DEFAULT";
  id: string;
  name: string;
}

export interface IfConditionNodePort {
  type: "IF";
  id: string;
  name: string;
  expression?: WorkflowValueDescriptor;
}

export interface ElifConditionNodePort {
  type: "ELIF";
  id: string;
  name: string;
  expression?: WorkflowValueDescriptor;
}

export interface ElseConditionNodePort {
  type: "ELSE";
  id: string;
  name: string;
}

export type NodePort =
  | DefaultNodePort
  | IfConditionNodePort
  | ElifConditionNodePort
  | ElseConditionNodePort;

export interface BaseWorkflowNode {
  type: string;
  definition?: CodeResourceDefinition;
}

export interface BaseDisplayableWorkflowNode extends BaseWorkflowNode {
  id: string;
  inputs: NodeInput[];
  type: string;
  displayData?: NodeDisplayData;
  base?: CodeResourceDefinition;
}

export interface EntrypointNodeData {
  label: string;
  sourceHandleId: string;
}

export interface EntrypointNode extends BaseDisplayableWorkflowNode {
  type: "ENTRYPOINT";
  data: EntrypointNodeData;
}

export interface CacheConfig {
  type?: "EPHEMERAL";
}

export interface JinjaPromptTemplateBlock {
  id: string;
  blockType: "JINJA";
  state: PromptBlockState;
  cacheConfig?: CacheConfig;
  properties: {
    template: string;
    templateType?: VellumVariableType;
  };
}

export interface ChatMessagePromptTemplateBlock {
  id: string;
  blockType: "CHAT_MESSAGE";
  state: PromptBlockState;
  cacheConfig?: CacheConfig;
  properties: {
    chatRole: ChatMessageRole;
    chatSource?: string;
    chatMessageUnterminated: boolean;
    blocks: PromptTemplateBlock[];
  };
}

export interface VariablePromptTemplateBlock {
  id: string;
  blockType: "VARIABLE";
  state: PromptBlockState;
  cacheConfig?: CacheConfig;
  inputVariableId: string;
}

export interface PlainTextPromptTemplateBlock {
  id: string;
  blockType: "PLAIN_TEXT";
  state: PromptBlockState;
  cacheConfig?: CacheConfig;
  text: string;
}

export type RichTextChildPromptTemplateBlock =
  | PlainTextPromptTemplateBlock
  | VariablePromptTemplateBlock;

export interface RichTextPromptTemplateBlock {
  id: string;
  blockType: "RICH_TEXT";
  state: PromptBlockState;
  cacheConfig?: CacheConfig;
  blocks: RichTextChildPromptTemplateBlock[];
}

export interface FunctionDefinitionPromptTemplateBlock {
  id: string;
  blockType: "FUNCTION_DEFINITION";
  state: PromptBlockState;
  cacheConfig?: CacheConfig;
  properties: {
    functionName?: string;
    functionDescription?: string;
    functionParameters?: Record<string, unknown>;
    functionForced?: boolean;
    functionStrict?: boolean;
  };
}

export type PromptTemplateBlock =
  | JinjaPromptTemplateBlock
  | ChatMessagePromptTemplateBlock
  | VariablePromptTemplateBlock
  | RichTextPromptTemplateBlock
  | FunctionDefinitionPromptTemplateBlock;

export interface PromptTemplateBlockData {
  version: number;
  blocks: PromptTemplateBlock[];
}

export interface PromptSettings {
  timeout?: number;
}

export interface PromptVersionExecConfig {
  parameters: PromptParameters;
  inputVariables: VellumVariable[];
  promptTemplateBlockData: PromptTemplateBlockData;
  settings?: PromptSettings;
}

export interface BasePromptNodeData {
  label: string;
  outputId: string;
  arrayOutputId: string;
  errorOutputId?: string;
  sourceHandleId: string;
  targetHandleId: string;
}

export interface InlinePromptNodeData extends BasePromptNodeData {
  variant: "INLINE";
  execConfig: PromptVersionExecConfig;
  mlModelName: string;
}

export interface DeploymentPromptNodeData extends BasePromptNodeData {
  variant: "DEPLOYMENT";
  promptDeploymentId: string;
  releaseTag: string;
  fallbackModels?: string[];
}

export interface PromptNodeSourceSandbox {
  sandboxId: string;
  promptId: string;
  sandboxSnapshotId: string;
}

export interface PromptVersionData {
  mlModelToWorkspaceId: string;
  execConfig: PromptVersionExecConfig;
}

export interface WorkflowSandboxRoutingConfig {
  version: number;
  promptVersionData?: PromptVersionData;
}

export interface PromptNodeDeployment {
  deploymentId: string;
  deploymentReleaseTagId: string;
}

export interface LegacyPromptNodeData extends BasePromptNodeData {
  variant: "LEGACY";
  sandboxRoutingConfig: WorkflowSandboxRoutingConfig;
  sourceSandbox?: PromptNodeSourceSandbox;
  deployment?: PromptNodeDeployment;
}

export type PromptNodeData =
  | InlinePromptNodeData
  | DeploymentPromptNodeData
  | LegacyPromptNodeData;

export interface InlinePromptNode extends BaseDisplayableWorkflowNode {
  type: "PROMPT";
  data: InlinePromptNodeData;
}

export interface PromptNode extends BaseDisplayableWorkflowNode {
  type: "PROMPT";
  data: PromptNodeData;
}

export interface SearchNodeData {
  label: string;
  resultsOutputId: string;
  textOutputId: string;
  errorOutputId?: string;
  sourceHandleId: string;
  targetHandleId: string;
  queryNodeInputId: string;
  documentIndexNodeInputId: string;
  weightsNodeInputId: string;
  limitNodeInputId: string;
  separatorNodeInputId: string;
  resultMergingEnabledNodeInputId: string;
  externalIdFiltersNodeInputId: string;
  metadataFiltersNodeInputId: string;
}

export interface SearchNode extends BaseDisplayableWorkflowNode {
  type: "SEARCH";
  data: SearchNodeData;
}

export interface DeploymentSubworkflowNodeData {
  label: string;
  sourceHandleId: string;
  targetHandleId: string;
  errorOutputId?: string;
  variant: "DEPLOYMENT";
  workflowDeploymentId: string;
  releaseTag: string;
}
export interface InlineSubworkflowNodeData {
  workflowRawData: WorkflowRawData;
  inputVariables: VellumVariable[];
  outputVariables: VellumVariable[];
  label: string;
  sourceHandleId: string;
  targetHandleId: string;
  errorOutputId?: string;
  variant: "INLINE";
}
export type SubworkflowNodeData =
  | ({
      variant: "DEPLOYMENT";
    } & DeploymentSubworkflowNodeData)
  | ({
      variant: "INLINE";
    } & InlineSubworkflowNodeData);
export interface SubworkflowNode extends BaseDisplayableWorkflowNode {
  type: "SUBWORKFLOW";
  data: SubworkflowNodeData;
}

export interface DeploymentMapNodeData {
  variant: "DEPLOYMENT";
  concurrency?: number;
  label: string;
  sourceHandleId: string;
  targetHandleId: string;
  errorOutputId?: string;
  itemsInputId: string;
  itemInputId: string;
  indexInputId: string;
  workflowDeploymentId: string;
  releaseTag: string;
}
export interface InlineMapNodeData {
  variant: "INLINE";
  workflowRawData: WorkflowRawData;
  inputVariables: VellumVariable[];
  outputVariables: VellumVariable[];
  concurrency?: number;
  label: string;
  sourceHandleId: string;
  targetHandleId: string;
  errorOutputId?: string;
  itemsInputId: string;
  itemInputId: string;
  indexInputId: string;
}
export type MapNodeData =
  | ({
      variant: "DEPLOYMENT";
    } & DeploymentMapNodeData)
  | ({
      variant: "INLINE";
    } & InlineMapNodeData);

export interface MapNode extends BaseDisplayableWorkflowNode {
  type: "MAP";
  data: MapNodeData;
}

export interface GuardrailNodeData {
  label: string;
  sourceHandleId: string;
  targetHandleId: string;
  errorOutputId?: string;
  metricDefinitionId: string;
  releaseTag: string;
}
export interface GuardrailNode extends BaseDisplayableWorkflowNode {
  type: "METRIC";
  data: GuardrailNodeData;
}

export type CodeExecutionPackage = {
  version: string;
  name: string;
};

export interface CodeExecutionNodeData {
  label: string;
  outputId: string;
  errorOutputId?: string;
  logOutputId?: string;
  sourceHandleId: string;
  targetHandleId: string;
  codeInputId: string;
  runtimeInputId: string;
  outputType: VellumVariableType;
  packages?: CodeExecutionPackage[];
  filepath?: string | null;
}
export interface CodeExecutionNode extends BaseDisplayableWorkflowNode {
  type: "CODE_EXECUTION";
  data: CodeExecutionNodeData;
}

export interface TemplatingNodeData {
  label: string;
  outputId: string;
  errorOutputId?: string;
  sourceHandleId: string;
  targetHandleId: string;
  templateNodeInputId: string;
  outputType: VellumVariableType;
}

export interface TemplatingNode extends BaseDisplayableWorkflowNode {
  type: "TEMPLATING";
  data: TemplatingNodeData;
}

export interface ConditionalRuleData {
  id: string;
  rules?: ConditionalRuleData[];
  combinator?: string;
  negated?: boolean;
  fieldNodeInputId?: string;
  operator?: string;
  valueNodeInputId?: string;
}

export interface ConditionalNodeConditionData {
  id: string;
  type: string;
  sourceHandleId: string;
  data?: ConditionalRuleData;
}

export interface ConditionalNodeData {
  label: string;
  targetHandleId: string;
  conditions: ConditionalNodeConditionData[];
  version: string;
}

export interface ConditionalNode extends BaseDisplayableWorkflowNode {
  type: "CONDITIONAL";
  data: ConditionalNodeData;
}

export interface FinalOutputNodeData {
  label: string;
  name: string;
  targetHandleId: string;
  outputId: string;
  outputType: VellumVariableType;
  nodeInputId: string;
}

export interface FinalOutputNode extends BaseDisplayableWorkflowNode {
  type: "TERMINAL";
  data: FinalOutputNodeData;
}

export interface MergeNodeTargetHandle {
  id: string;
}

export interface MergeNodeData {
  label: string;
  mergeStrategy: "AWAIT_ALL" | "AWAIT_ANY";
  targetHandles: MergeNodeTargetHandle[];
  sourceHandleId: string;
}

export interface MergeNode extends BaseDisplayableWorkflowNode {
  type: "MERGE";
  data: MergeNodeData;
}

export interface ApiNodeAdditionalHeaderData {
  headerKeyInputId: string;
  headerValueInputId: string;
}

export interface ApiNodeData {
  label: string;
  methodInputId: string;
  urlInputId: string;
  bodyInputId: string;
  authorizationTypeInputId?: string;
  bearerTokenValueInputId?: string;
  apiKeyHeaderKeyInputId?: string;
  apiKeyHeaderValueInputId?: string;
  additionalHeaders?: ApiNodeAdditionalHeaderData[];
  textOutputId: string;
  jsonOutputId: string;
  statusCodeOutputId: string;
  errorOutputId?: string;
  targetHandleId: string;
  sourceHandleId: string;
}

export interface ApiNode extends BaseDisplayableWorkflowNode {
  type: "API";
  data: ApiNodeData;
}

export interface NoteNodeData {
  label: string;
  text?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  style?: Record<string, any>;
}

export interface NoteNode extends BaseDisplayableWorkflowNode {
  type: "NOTE";
  data: NoteNodeData;
}

export interface ErrorNodeData {
  label: string;
  name: string;
  targetHandleId: string;
  errorSourceInputId: string;
  errorOutputId: string;
}

export interface ErrorNode extends BaseDisplayableWorkflowNode {
  type: "ERROR";
  data: ErrorNodeData;
}

export interface GenericNodeDisplayData {
  position?: { x: number; y: number };
}

export interface NodeOutput {
  id: string;
  name: string;
  type: VellumVariableType;
  value?: WorkflowValueDescriptor;
}

export interface GenericNode extends BaseWorkflowNode {
  id: string;
  type: "GENERIC";
  displayData?: GenericNodeDisplayData;
  base: CodeResourceDefinition;
  definition?: CodeResourceDefinition;
  trigger: NodeTrigger;
  ports: NodePort[];
  adornments?: AdornmentNode[];
  attributes: NodeAttribute[];
  outputs: NodeOutput[];
}

export type WorkflowDataNode =
  | PromptNode
  | SearchNode
  | SubworkflowNode
  | TemplatingNode
  | MapNode
  | GuardrailNode
  | CodeExecutionNode
  | FinalOutputNode
  | MergeNode
  | ConditionalNode
  | ApiNode
  | NoteNode
  | ErrorNode
  | GenericNode;

export type WorkflowNode = WorkflowDataNode | EntrypointNode | FinalOutputNode;

export interface WorkflowEdge {
  id: string;
  type: "DEFAULT";
  sourceNodeId: string;
  sourceHandleId: string;
  targetNodeId: string;
  targetHandleId: string;
}

export interface WorkflowDisplayDataViewport {
  x: number;
  y: number;
  zoom: number;
}

export interface WorkflowDisplayData {
  viewport: WorkflowDisplayDataViewport;
}

export interface WorkflowRawData {
  nodes: WorkflowNode[];
  edges: WorkflowEdge[];
  displayData?: WorkflowDisplayData;
  definition?: CodeResourceDefinition;
}

export interface RunnerConfig {
  containerImageName?: string;
  containerImageTag?: string;
}

export interface WorkflowVersionExecConfig {
  workflowRawData: WorkflowRawData;
  inputVariables: VellumVariable[];
  outputVariables: VellumVariable[];
  runnerConfig?: RunnerConfig;
}

type WorkflowSandboxInput =
  | StringInput
  | JsonInput
  | ChatHistoryInput
  | NumberInput;
export type WorkflowSandboxInputs = WorkflowSandboxInput[];

export interface UnaryWorkflowExpression {
  type: "UNARY_EXPRESSION";
  lhs: WorkflowValueDescriptor;
  operator: LogicalOperator;
}

export interface BinaryWorkflowExpression {
  type: "BINARY_EXPRESSION";
  lhs: WorkflowValueDescriptor;
  operator: LogicalOperator;
  rhs: WorkflowValueDescriptor;
}

export interface TernaryWorkflowExpression {
  type: "TERNARY_EXPRESSION";
  base: WorkflowValueDescriptor;
  operator: LogicalOperator;
  lhs: WorkflowValueDescriptor;
  rhs: WorkflowValueDescriptor;
}

export interface NodeOutputWorkflowReference {
  type: "NODE_OUTPUT";
  nodeId: string;
  nodeOutputId: string;
}

export interface WorkflowInputWorkflowReference {
  type: "WORKFLOW_INPUT";
  inputVariableId: string;
}

export interface WorkflowStateVariableWorkflowReference {
  type: "WORKFLOW_STATE";
  stateVariableId: string;
}

export interface ConstantValueWorkflowReference {
  type: "CONSTANT_VALUE";
  value: VellumValue;
}

export interface VellumSecretWorkflowReference {
  type: "VELLUM_SECRET";
  vellumSecretName: string;
}

export interface ExecutionCounterWorkflowReference {
  type: "EXECUTION_COUNTER";
  nodeId: string;
}

export type WorkflowValueDescriptorReference =
  | NodeOutputWorkflowReference
  | WorkflowInputWorkflowReference
  | WorkflowStateVariableWorkflowReference
  | ConstantValueWorkflowReference
  | VellumSecretWorkflowReference
  | ExecutionCounterWorkflowReference;

export type WorkflowValueDescriptor =
  | UnaryWorkflowExpression
  | BinaryWorkflowExpression
  | TernaryWorkflowExpression
  | WorkflowValueDescriptorReference;

export interface NodeAttribute {
  id: string;
  name: string;
  value: WorkflowValueDescriptor;
}

export interface AdornmentNode {
  id: string;
  label: string;
  base: CodeResourceDefinition;
  attributes: NodeAttribute[];
}

export type OperatorMapping =
  | "equals"
  | "does_not_equal"
  | "less_than"
  | "greater_than"
  | "less_than_or_equal_to"
  | "greater_than_or_equal_to"
  | "contains"
  | "begins_with"
  | "ends_with"
  | "does_not_contain"
  | "does_not_begin_with"
  | "does_not_end_with"
  | "is_null"
  | "is_not_null"
  | "in"
  | "not_in"
  | "between"
  | "not_between";

export interface IterableConfig {
  endWithComma?: boolean;
}
