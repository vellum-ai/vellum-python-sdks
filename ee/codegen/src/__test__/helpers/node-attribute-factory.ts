import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
