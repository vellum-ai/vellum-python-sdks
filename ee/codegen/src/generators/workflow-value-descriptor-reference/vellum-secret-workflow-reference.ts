import { python } from "@fern-api/python-ast";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { isNil } from "lodash";

import { BaseNodeInputWorkflowReference } from "src/generators/workflow-value-descriptor-reference/BaseNodeInputWorkflowReference";
import { VellumSecretWorkflowReference as VellumSecretWorkflowReferenceType } from "src/types/vellum";

export class VellumSecretWorkflowReference extends BaseNodeInputWorkflowReference<VellumSecretWorkflowReferenceType> {
  getAstNode(): AstNode | undefined {
    const vellumSecretName =
      this.nodeInputWorkflowReferencePointer.vellumSecretName;

    if (isNil(vellumSecretName)) {
      return python.TypeInstantiation.none();
    }
    return python.instantiateClass({
      classReference: python.reference({
        name: "VellumSecretReference",
        modulePath: [
          ...this.workflowContext.sdkModulePathNames.WORKFLOWS_MODULE_PATH,
          "references",
        ],
      }),
      arguments_: [
        python.methodArgument({
          value: python.TypeInstantiation.str(vellumSecretName),
        }),
      ],
    });
  }
}
