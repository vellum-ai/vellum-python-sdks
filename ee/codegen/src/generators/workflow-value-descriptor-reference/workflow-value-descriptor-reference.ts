import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";

import { WorkflowContext } from "src/context";
import { ValueGenerationError } from "src/generators/errors";
import { BaseNodeInputWorkflowReference } from "src/generators/workflow-value-descriptor-reference/BaseNodeInputWorkflowReference";
import { ConstantValueReference } from "src/generators/workflow-value-descriptor-reference/constant-value-reference";
import { ExecutionCounterWorkflowReference } from "src/generators/workflow-value-descriptor-reference/execution-counter-workflow-reference";
import { NodeOutputWorkflowReference } from "src/generators/workflow-value-descriptor-reference/node-output-workflow-reference";
import { VellumSecretWorkflowReference } from "src/generators/workflow-value-descriptor-reference/vellum-secret-workflow-reference";
import { WorkflowInputReference } from "src/generators/workflow-value-descriptor-reference/workflow-input-reference";
import {
  IterableConfig,
  WorkflowValueDescriptorReference as WorkflowValueDescriptorReferenceType,
} from "src/types/vellum";
import { assertUnreachable } from "src/utils/typing";

export declare namespace WorkflowValueDescriptorReference {
  export interface Args {
    workflowContext: WorkflowContext;
    workflowValueReferencePointer: WorkflowValueDescriptorReferenceType;
    iterableConfig?: IterableConfig;
  }
}

export class WorkflowValueDescriptorReference extends AstNode {
  private workflowContext: WorkflowContext;
  public readonly workflowValueReferencePointer: WorkflowValueDescriptorReferenceType["type"];
  private iterableConfig?: IterableConfig;
  public astNode:
    | BaseNodeInputWorkflowReference<WorkflowValueDescriptorReferenceType>
    | undefined;

  constructor(args: WorkflowValueDescriptorReference.Args) {
    super();

    this.workflowContext = args.workflowContext;
    this.workflowValueReferencePointer =
      args.workflowValueReferencePointer.type;
    this.iterableConfig = args.iterableConfig;

    this.astNode = this.getAstNode(args.workflowValueReferencePointer);

    if (this.astNode) {
      this.inheritReferences(this.astNode);
    }
  }

  private getAstNode(
    workflowValueReferencePointer: WorkflowValueDescriptorReferenceType
  ):
    | BaseNodeInputWorkflowReference<WorkflowValueDescriptorReferenceType>
    | undefined {
    const referenceType = workflowValueReferencePointer.type;

    switch (referenceType) {
      case "NODE_OUTPUT": {
        const reference = new NodeOutputWorkflowReference({
          workflowContext: this.workflowContext,
          nodeInputWorkflowReferencePointer: workflowValueReferencePointer,
        });
        if (reference.getAstNode()) {
          return reference;
        } else {
          return undefined;
        }
      }
      case "WORKFLOW_INPUT":
        return new WorkflowInputReference({
          workflowContext: this.workflowContext,
          nodeInputWorkflowReferencePointer: workflowValueReferencePointer,
        });
      case "WORKFLOW_STATE":
        this.workflowContext.addError(
          new ValueGenerationError(
            "WORKFLOW_STATE reference pointers is not implemented"
          )
        );
        return undefined;
      case "CONSTANT_VALUE":
        return new ConstantValueReference({
          workflowContext: this.workflowContext,
          nodeInputWorkflowReferencePointer: workflowValueReferencePointer,
          iterableConfig: this.iterableConfig,
        });
      case "VELLUM_SECRET":
        return new VellumSecretWorkflowReference({
          workflowContext: this.workflowContext,
          nodeInputWorkflowReferencePointer: workflowValueReferencePointer,
        });
      case "EXECUTION_COUNTER":
        return new ExecutionCounterWorkflowReference({
          workflowContext: this.workflowContext,
          nodeInputWorkflowReferencePointer: workflowValueReferencePointer,
        });
      default: {
        assertUnreachable(referenceType);
      }
    }
  }

  public write(writer: Writer): void {
    if (this.astNode) {
      this.astNode.write(writer);
    }
  }
}
