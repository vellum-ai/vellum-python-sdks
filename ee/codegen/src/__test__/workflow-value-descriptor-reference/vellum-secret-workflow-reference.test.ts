import { Writer } from "@fern-api/python-ast/core/Writer";

import { workflowContextFactory } from "src/__test__/helpers";
import { WorkflowContext } from "src/context";
import { VellumSecretWorkflowReference } from "src/generators/workflow-value-descriptor-reference/vellum-secret-workflow-reference";
import { WorkflowValueDescriptorReference } from "src/types/vellum";

describe("VellumSecretWorkflowReferencePointer", () => {
  let workflowContext: WorkflowContext;

  beforeEach(() => {
    workflowContext = workflowContextFactory();
  });

  it("should generate correct AST for vellum secret reference", async () => {
    const secretReference: WorkflowValueDescriptorReference = {
      type: "VELLUM_SECRET",
      vellumSecretName: "test-secret",
    };

    const pointer = new VellumSecretWorkflowReference({
      workflowContext,
      nodeInputWorkflowReferencePointer: secretReference,
    });

    const writer = new Writer();
    pointer.write(writer);

    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });
});
