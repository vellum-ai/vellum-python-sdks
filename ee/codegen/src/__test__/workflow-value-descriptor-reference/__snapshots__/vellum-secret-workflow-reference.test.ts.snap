// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VellumSecretWorkflowReferencePointer > should generate correct AST for vellum secret reference 1`] = `
"VellumSecretReference("test-secret")
"
`;
